build_tools/packaging/python/templates/rocm/src/rocm_sdk/_devel.py

```diff
@@ -164,20 +167,47 @@ def _clean_dir(dir: Path):
                         tf.extract(ti, path=site_lib_path)
             tarfile_path.unlink()
         finally:
-            _unlock_file(record_file)
+            file_lock.unlock()
 
 
-def _lock_file(f):
-    # TODO: This is Posix specific file locking. We should fork on Windows to use
-    # msvcrt locking.
-    import fcntl, os
+def _is_windows():
+    return platform.system() == "Windows"
 
-    fcntl.lockf(f, fcntl.LOCK_EX)
 
+class FileLock:
+    """Small portability shim between fcntl.lockf and msvcrt.locking for our uses."""
 
-def _unlock_file(f):
-    # TODO: This is Posix specific file locking. We should fork on Windows to use
-    # msvcrt locking.
-    import fcntl, os
+    def __init__(self, file: io.TextIOWrapper):
+        self.file = file
+        self.original_file_size = os.path.getsize(file.name)
 
-    fcntl.lockf(f, fcntl.LOCK_UN)
+        if _is_windows():
+            # The Windows APIs for file locking apply to only a given range
+            # within the file and lock/unlock calls must be balanced. Since we
+            # will be appending to the locked file, we lock as much as we know
+            # about (the 'nbytes' parameter can continue beyond the end of the
+            # file, but we don't know how much we'll be writing ahead of time).
```

Comment:
Seems like it should have the desired effect. Since any two calls will be locking overlapping subsets of the range, it should achieve mutual exclusion.